### ISSUE TITLE
Grouping does not work for non-BNet friends again

### ISSUE THREAD

**Issue by jfarmer** [Mikeprod/FriendGroups#35]
The commits adding `C_FriendList.GetFriendInfoByIndex` break the addon for non-BNET friends (e.g., https://github.com/Mikeprod/FriendGroups/commit/f3b6b065ca0a122be4adac11d010386e67bc049f).

Per https://wow.gamepedia.com/API_C_FriendList.GetFriendInfo, `C_FriendList.GetFriendInfoByIndex` returns a table. `select` works on argument lists, not tables.  Use `/dump` on the following expressions to see what's going on:

```lua
select(2, "hey", "there", "friend") -- every argument from index 2 onward
select(2, {"hey", "there", "friend"}) -- empty result
select(2, unpack({"hey", "there", "friend"})) -- every item in the table from index 2 onward
```

The code is written as if `select(2, {"hey", "there", "friend"})` makes sense.

Fortunately (again, per the documentation plus quick testing in game), `C_FriendList.GetFriendInfoByIndex` returns a table with named keys unlike most of the WoW API.

This means there's no need to assign the fields to `local` variables and instead, we can do something like this:

```lua
local friendInfo = C_FriendList.GetFriendInfoByIndex(j)
friendInfo.note
friendInfo.className
```

**Comment by Mikeprod** [comment]
Same old same old, I do not remember that the API function returns a table and not several values as the previous functions ... I will fix it, AGAIN ! :) Thanks for reporting

**Comment by jfarmer** [comment]
One (engineering) process suggestion would be to create one or more test characters with friend lists in various states. Before shipping a new version you log in and verify that their friend lists look as expected. Each test list should be simple enough that you can verify whether it is correct at a glance.

Some test cases:

- A group containing only BNET friends
- A group containing only non-BNET friends
- A group containing a mix of both BNET/non-BNET friends

I'd also consider creating a debug mode via a cvar or whatever's appripriate.

You could create special debug-mode-only flags that you embed in a user note to force various states. This would be a good way to test that things render as expected when a friend is in a specific state without a friend having to actually be in a particular state.

For example, if you want to test that non-bnet friends render correctly when online vs. offline. The note might be something like `!onlineState:offline # Offline Test Group`. When the aforementioned debug mode is enabled, it'd parse `!onlineState` to override the WOW API-supplied values, rendering the friend as if they're offline regardless of their actual state.  It'd then render the `Offline Test Group` as normal, oblivious to the fact that the underlying user's online status has been overridden.

Values might be `online`, `afk`, and `offline`.

I'd start simple and focus on making as painless as possible to detect major regressions related to rendering non-BNET friends, 3-5 test scenarios max.

**Comment by Mikeprod** [comment]
Yes, I agree. Having test would be something big to ensure the addon stability. The volatility of a friend list is high, therefore setting up use cases would be a great deal as you mentionned :)
